cosmos/launchpad/block.go

```diff
@@ -0,0 +1,180 @@
+package launchpad
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/antihax/optional"
+	"github.com/coinbase/rosetta-sdk-go/types"
+	cosmostypes "github.com/cosmos/cosmos-sdk/types"
+	tendermintclient "github.com/tendermint/cosmos-rosetta-gateway/cosmos/launchpad/client/tendermint/generated"
+	"github.com/tendermint/cosmos-rosetta-gateway/pkg/ensurehex"
+)
+
+func (l Launchpad) Block(ctx context.Context, r *types.BlockRequest) (*types.BlockResponse, *types.Error) {
+	var (
+		blockResp       tendermintclient.BlockResponse
+		parentBlockResp tendermintclient.BlockResponse
+		err             error
+	)
+	if r.BlockIdentifier.Index != nil {
+		blockResp, _, err = l.tendermint.Info.Block(ctx, &tendermintclient.BlockOpts{
+			Height: optional.NewFloat32(float32(*r.BlockIdentifier.Index)),
+		})
+	} else {
+		blockResp, _, err = l.tendermint.Info.BlockByHash(ctx, ensurehex.String(*r.BlockIdentifier.Hash))
+	}
+	if err != nil {
+		return nil, ErrNodeConnection
+	}
+
+	timestamp, err := time.Parse(time.RFC3339Nano, blockResp.Result.Block.Header.Time)
+	if err != nil {
+		return nil, ErrInterpreting
+	}
+
+	txsquery := fmt.Sprintf(`"tx.height=%s"`, blockResp.Result.Block.Header.Height)
+	txsResp, _, err := l.tendermint.Info.TxSearch(ctx, txsquery, nil)
+	if err != nil {
+		return nil, ErrNodeConnection
+	}
+	transactions, err := toTransactions(txsResp.Result)
+	if err != nil {
+		return nil, ErrInterpreting
+	}
+
+	block, err := toBlockIdentifier(blockResp.Result)
+	if err != nil {
+		return nil, ErrInterpreting
+	}
+
+	resp := &types.BlockResponse{
+		Block: &types.Block{
+			BlockIdentifier: block,
+			Transactions:    transactions,
+			Timestamp:       timestamp.UnixNano() / 1000000,
+		},
+	}
+	hasParentBlock := blockResp.Result.Block.Header.LastBlockId.Hash != ""
+	if hasParentBlock {
+		parentBlockResp, _, err = l.tendermint.Info.BlockByHash(ctx, ensurehex.String(blockResp.Result.Block.Header.LastBlockId.Hash))
+		if err != nil {
+			return nil, ErrNodeConnection
+		}
+		parentBlock, err := toBlockIdentifier(parentBlockResp.Result)
+		if err != nil {
+			return nil, ErrInterpreting
+		}
+		resp.Block.ParentBlockIdentifier = parentBlock
+	}
+	return resp, nil
+}
+
+func (l Launchpad) BlockTransaction(ctx context.Context, r *types.BlockTransactionRequest) (*types.BlockTransactionResponse, *types.Error) {
+	panic("unimplemented")
+}
+
+func toBlockIdentifier(result tendermintclient.BlockComplete) (*types.BlockIdentifier, error) {
+	if result.BlockId.Hash == "" {
+		return nil, nil
+	}
+	height, err := strconv.ParseUint(result.Block.Header.Height, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	return &types.BlockIdentifier{
+		Index: int64(height),
+		Hash:  result.BlockId.Hash,
+	}, nil
+}
+
+func toTransactions(result tendermintclient.TxSearchResponseResult) (transactions []*types.Transaction, err error) {
+	for _, tx := range result.Txs {
+		var operations []*types.Operation
```

Comment:
@jgimeno how can we improve mapping Tendermint's TX Events to Rosetta Operations? Can you please check the code below.

Basically I need more info on how to map Events with `message` types to Operations.

Reply:
Btw, I'll simplify this func (toTransactions) after your check on mapping

Reply:
checking

Reply:
Ok this needs a little bit of investigation by itself, we need to collect all types of transactions that can be performed on the cosmos SDK launchpad. 

After that we need to tag them to a specific name. And after it we need to learn how this data is represented in the request, because not all transactions throw Events but does not mean it is not a specific transaction. This last part I can do it, provide the information by transaction and transaction type and how to extract it from the data you receive. (Unless is something you would like to do).

Reply:
implemented what we have came up with in our call https://github.com/tendermint/cosmos-rosetta-gateway/pull/34/commits/a7aef1275c11f63fc07ef7f8c70e083273922dad